Fix saving belongstomany with complex key values

markstory opened the pull request:
Fix incorrectly persisted data when a belongsToMany association foreignKey/bindingKey fields were modeled as complex types like datetime instances.

The strict comparison on the entire key data needed to be replaced with a more nuanced solution. At worst this code will operate in linear time, but I think that N will be <2 for the vast majority of applications. I've attempted to make the solution as efficient as possible and failures should only make one loop iteration.

Fixes #16562

markstory committed:
Fix saving belongstomany with complex key values

Fix incorrectly persisted data when a belongsToMany association
foreignKey/bindingKey fields were modeled as complex types like datetime
instances.

The strict comparison on the entire key data needed to be replaced
with a more nuanced solution. At worst this code will operate in linear
time, but I think that N will be <2 for the vast majority of
applications. I've attempted to make the solution as efficient as
possible and failures should only make one loop iteration.

Fixes #16562

othercorey commented:
Think we need 100% coverage on this. There is one case missing according to codecov.

othercorey commented:
This code could benefit from some naming cleanup. It has variables named `$result`, `$indexed` and `$data` together in a loop.

```php
        foreach ($existing as $result) {
            $fields = $result->extract($keys);
            $found = false;
            foreach ($indexed as $i => $data) {
 ```

ADmad commented:
PHP 8.2 has a new deprecation for use of `utf8_decode()`.

markstory commented:
> PHP 8.2 has a new deprecation for use of utf8_decode().

Well that's annoying :cry:

markstory committed:
Improve variable naming.

markstory committed:
Remove usage of deprecated function.

othercorey reviewed (approved):
The logic looks ok to me. Just needs a test case for the array key not set condition.

ADmad commented:
@markstory The deprecation error has already been fixed on 4.x branch.

markstory committed:
Merge branch '4.x' into issue-16562

markstory committed:
Remove redundant case

Both arrays will always have all keys because of how
EntityTrait::extract() works.
